modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,707 @@
+                yield obj
+
+
+def _construct_list(
+    lst: List,
+    chain: List,
+    refs: Dict[int, Any],
+    meta: List,
+):  # pragma: no cover # noqa: GL08
+    pop = chain.pop
+    lst_append = lst.append
+    while True:
+        obj = pop()
+        if isinstance(obj, _Tag):
+            if obj == _Tag.END:
+                break
+            elif obj == _Tag.CHAIN:
+                yield _construct_chain(chain, refs, meta, lst)
+            elif obj == _Tag.LIST:
+                lst_append([])
+                yield _construct_list(lst[-1], chain, refs, meta)
+            elif obj is _Tag.REF:
+                lst_append(refs[pop()])
+            else:
+                raise ValueError(f"Unexpected tag {obj}")
+        else:
+            lst_append(obj)
+
+
+@ray.remote(num_returns=4)
+def remote_exec_func(
+    fn: Callable, obj: Any, *args: Tuple, **kwargs: Dict
+):  # pragma: no cover
+    """
+    Execute the specified function in a worker process.
+
+    The object `obj` is passed to the function as the first argument of.
+
+    Parameters
+    ----------
+    fn : Callable
+    obj : Any
+    *args : list
+    **kwargs : dict
+
+    Returns
+    -------
+    tuple[Any, int, int, str]
+    The execution result, the result length and width, the worked address.
+    """
+    try:
+        obj = _exec_func(fn, obj, args, kwargs)
+        return (
+            obj,
+            len(obj) if hasattr(obj, "__len__") else 0,
+            len(obj.columns) if hasattr(obj, "columns") else 0,
+            get_node_ip_address(),
+        )
+    except DeferredExecutionException as err:
+        return [err] * 4
+
+
+def _exec_func(fn: Callable, obj: Any, args: Tuple, kwargs: Dict):  # noqa: GL08
+    try:
+        try:
+            return fn(obj, *args, **kwargs)
+            # Sometimes Arrow forces us to make a copy of an object before we operate on it. We
+            # don't want the error to propagate to the user, and we want to avoid copying unless
+            # we absolutely have to.
+        except ValueError as err:
+            if isinstance(obj, (pandas.DataFrame, pandas.Series)):
+                return fn(obj.copy(), *args, **kwargs)
+            else:
+                raise err
+    except Exception as err:
+        raise DeferredExecutionException(
+            f"fn={fn}, obj={obj}, args={args}, kwargs={kwargs}", err
+        )
+
+
+def _remote_exec_chain(num_returns: int, *args: Tuple):
+    """
+    Execute the deconstructed chain in a worker process.
+
+    Parameters
+    ----------
+    num_returns : int
+        The number of return values.
+    *args : tuple
+        A deconstructed chain to be executed.
+
+    Returns
+    -------
+    Generator
+    """
+    if num_returns == 2:
+        return _remote_exec_single_chain.remote(*args)
+    else:
+        return _remote_exec_multi_chain.options(num_returns=num_returns).remote(
+            num_returns, *args
+        )
```

Comment:
Why do we need this branching?

Reply:
A minor optimisation for a frequent case with a single return value. The num_returns is specified statically, but not in options() on each call. Besides, the num_returns argument is not passed to the remote function.

Reply:
Where do you see the benefit from this branching?

Reply:
1. We save on 1 argument passed to remote function.
2. According to my tests, if num_returns is specified directly in @ray.remote(num_returns=2), it works faster than doing it dynamically with .options(num_returns=num_returns).